docs/docs/remark-ast.md

```diff
@@ -0,0 +1,201 @@
+---
+title: Manipulating Remark ASTs
+---
+
+gatsby-transformer-remark empowers developers to translate Markdown into HTML to be consumed throughout Gatsby pages. Blogs and other content based sites can highly benefit from this so the code and content is separated. With this, people who will write content for the site don't need to worry about how the the site was written, but rather can focus just on writing the Markdown.
+
+In certain instances, a developer may want to customize certain Markdown nodes while being parsed. Examples could include [add syntax highlighting](/packages/gatsby-remark-prismjs/), [parse images](/packages/gatsby-remark-images), [embed videos](/packages/gatsby-remark-embed-video) and plenty of others. Throughout all of these instances, a plugin will examine the Markdown Abstract Syntax Tree (AST) and manipulate content based on certain node types or content in particular nodes.
+
+## What you will learn in this tutorial
+
+- How to learn how to understand the remark AST
+- How to integrate plugins with `gatsby-transformer-remark`
+- How to manipulate the remark AST to add additional functionality
+
+## Understanding the Abstract Syntax Tree
+
+To get an understanding at what is available in the Markdown AST, take a look at the markdown ast spec that is used in remark and other unist projects: [syntax-tree/mdast](https://github.com/syntax-tree/mdast).
+
+Starting out with a Markdown file as below:
+
+```markdown
+# Hello World!
+
+This is a [Real page](https://google.com)
+```
+
+Remark would translate this into an AST that will be available to `gatsby-transformer-remark` plugins. [AST Explorer](https://astexplorer.net/#/gist/d9029a2e8827265fbb9b190083b59d4d/3384f3ce6a3084e50043d0c8ce34628ed7477603) is a site that gives you a side-by-side view of the markdown and the outputted AST.
+
+## Setting up a plugin
+
+You are going to create a plugin that colors all top-level headings in the markdown with the color purple.
+
+First create a local plugin by adding a `plugins` folder in your site and generating a package.json file for it. As well, create an index.js file. In this file, it will export a single function.
+
+```js:title=plugins/gatsby-remark-purple-headers/index.js
+module.exports = ({ markdownAST }, pluginOptions) => {
+  // Manipulate AST
+
+  return markdownAST
+}
+```
+
+The first parameter is all of the default properties that can be used in plugins (actions, store, getNodes, schema, etc.) plus a couple just for gatsby-transformer-remark plugins. The one to be focused on is the `markdownAST` field which is destructured in the code snippet above.
+
+As with other Gatsby plugins, the 2nd parameter is the `pluginOptions` which is obtained from the definition in `gatsby-config.js` file.
+
+Finally, the function will return the `markdownAST` after the fields you wish to be edited are transformed.
```

Comment:
Doesn't have to since it's mutated--but a nit ;)

Reply:
Also I like showing best practices here--so even though we're mutating, returning is a best practice for e.g. easy authoring of unit tests.